src/server/game/Entities/Player/Player.cpp

```diff
@@ -14960,15 +14960,69 @@ void Player::RewardQuest(Quest const* quest, LootItemType rewardType, uint32 rew
         UpdatePvPState();
     }
 
-    SendQuestUpdate(quest_id);
-
     SendQuestGiverStatusMultiple();
 
+    SendQuestUpdate(quest_id, false);
+
     //lets remove flag for delayed teleports
     SetCanDelayTeleport(false);
 
+    bool canHaveNextQuest = !quest->HasFlag(QUEST_FLAGS_AUTOCOMPLETE) ? !questGiver->IsPlayer() : true;
+    if (canHaveNextQuest)
+    {
+        switch (questGiver->GetTypeId())
+        {
+            case TYPEID_UNIT:
+            case TYPEID_PLAYER:
+            {
+                //For AutoSubmition was added plr case there as it almost same exclute AI script cases.
+                // Send next quest
+                if (Quest const* nextQuest = GetNextQuest(questGiver->GetGUID(), quest))
+                {
+                    // Only send the quest to the player if the conditions are met
+                    if (CanTakeQuest(nextQuest, false))
+                    {
+                        if (nextQuest->IsAutoAccept() && CanAddQuest(nextQuest, true))
+                            AddQuestAndCheckCompletion(nextQuest, questGiver);
+
+                        PlayerTalkClass->SendQuestGiverQuestDetails(nextQuest, questGiver->GetGUID(), true, false);
+                    }
+                }
+
+                PlayerTalkClass->ClearMenus();
+                if (Creature* creatureQGiver = questGiver->ToCreature())
+                    creatureQGiver->AI()->OnQuestReward(this, quest, rewardType, rewardId);
+                break;
+            }
+            case TYPEID_GAMEOBJECT:
+            {
+                GameObject* questGiverGob = questGiver->ToGameObject();
+                // Send next quest
+                if (Quest const* nextQuest = GetNextQuest(questGiverGob->GetGUID(), quest))
+                {
+                    // Only send the quest to the player if the conditions are met
+                    if (CanTakeQuest(nextQuest, false))
+                    {
+                        if (nextQuest->IsAutoAccept() && CanAddQuest(nextQuest, true))
+                            AddQuestAndCheckCompletion(nextQuest, questGiver);
+
+                        PlayerTalkClass->SendQuestGiverQuestDetails(nextQuest, questGiverGob->GetGUID(), true, false);
+                    }
+                }
+
+                PlayerTalkClass->ClearMenus();
+                questGiverGob->AI()->OnQuestReward(this, quest, rewardType, rewardId);
+                break;
+            }
+            default:
+                break;
+        }
+    }
+
     sScriptMgr->OnQuestStatusChange(this, quest_id);
     sScriptMgr->OnQuestStatusChange(this, quest, oldStatus, QUEST_STATUS_REWARDED);
+
+    UpdateObjectVisibility();
```

Comment:
and where does this come from?

Reply:
Like you see above, I purposefully set false for SendQuestUpdate, so the phases are sent if needed but visibility is updated at a later point after the AI hooks are done

Reply:
reread 2nd sentence in my comment, you changed "update visibility if phases changed earlier" to "update visibility always but later"

Reply:
Ah yes that’s what I do, otherwise I would have to return a Boolean in OnConditionChanged so I know if a phase changed, the question is: is it worth it ?